src/main/scala/firrtl/DiGraph.scala

```diff
@@ -0,0 +1,158 @@
+package firrtl
+
+import scala.collection.immutable.{HashSet, HashMap}
+import scala.collection.mutable
+import scala.collection.mutable.MultiMap
+
+class MutableDiGraph[T](
+  val edgeData: MultiMap[T,T] = new mutable.HashMap[T, mutable.Set[T]] with MultiMap[T, T]) {
+  def contains(v: T) = edgeData.contains(v)
+  def getVertices = edgeData.keys
+  def getEdges(v: T) = edgeData(v)
+  def addVertex(v: T): T = {
+    edgeData.getOrElseUpdate(v,new mutable.HashSet[T])
+    v
+  }
+  // Add v to keys to maintain invariant
+  def addEdge(u: T, v: T) = {
+    edgeData.getOrElseUpdate(v, new mutable.HashSet[T])
+    edgeData.addBinding(u,v)
+  }
+}
+
+object DiGraph {
+  def apply[T](mdg: MutableDiGraph[T]) = new DiGraph((mdg.edgeData mapValues { _.toSet }).toMap[T, Set[T]])
+  def apply[T](edgeData: MultiMap[T,T]) = new DiGraph((edgeData mapValues { _.toSet }).toMap[T, Set[T]])
+}
+
+class DiGraph[T] (val edges: Map[T, Set[T]]) {
+
+  def getVertices = edges.keys
+  def getEdges(v: T) = edges.getOrElse(v, new HashSet[T])
+
+  // Graph must be acyclic for valid linearization
+  def linearize(root: T) = {
```

Comment:
If the graph is cyclic and I call linearize, what does your function do? This should be explained in the comment. Ideally, you should error if the graph is cyclic.